apiml/src/main/java/org/zowe/apiml/controller/ReactiveSafResourceAccessController.java

```
@@ -0,0 +1,78 @@
+import lombok.NoArgsConstructor;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatusCode;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.core.context.ReactiveSecurityContextHolder;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import org.zowe.apiml.message.core.MessageService;
+import org.zowe.apiml.security.common.auth.saf.AccessLevel;
+import org.zowe.apiml.security.common.auth.saf.SafResourceAccessVerifying;
+import reactor.core.publisher.Mono;
+
+@RestController
+@RequestMapping("/gateway/auth")
+@RequiredArgsConstructor
+@Slf4j
+public class ReactiveSafResourceAccessController {
+
+    private final SafResourceAccessVerifying safResourceAccessVerifying;
+    private final MessageService messageService;
+
+    @PostMapping(path = "/check", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public Mono<ResponseEntity<?>> hasSafAccess(@RequestBody CheckRequestModel request) {
+        return ReactiveSecurityContextHolder.getContext()
+            .flatMap(ctx -> Mono.justOrEmpty(ctx.getAuthentication()))
+            .filter(auth -> auth.getPrincipal() != null)
+            .flatMap(authentication -> {
+                if (safResourceAccessVerifying.hasSafResourceAccess(
+                    authentication,
+                    request.getResourceClass(),
+                    request.getResourceName(),
+                    request.getAccessLevel().name()
+                )) {
+                    return Mono.just(ResponseEntity.noContent().build());
+                } else {
+                    log.debug("Access denied for user: {}", authentication.getPrincipal());
```

We should throw an authentication exception here. The error is the same as if authentication is not provided...